File: ProtonMail/ProtonMail/LanguageManager.m
//
//  LocalizationManager.m
//  ProtonMail
//
//  Created by Yanfeng Zhang on 6/5/17.
//  Copyright © 2017 ProtonMail. All rights reserved.
//


#import "LanguageManager.h"
#import "NSBundle+Language.h"

static NSString * const LanguageCodes[] = { @"en", @"de", @"fr",
                                            @"ru", @"es", @"tr",
                                            @"pl", @"uk", @"nl",
                                            @"it", @"pt-BR"
};

static NSString * const LanguageStrings[] = { @"English", @"German", @"French",
                                              @"Russian", @"Spanish", @"Turkish",
                                              @"Polish", @"Ukrainian", @"Dutch",
                                              @"Italian", @"PortugueseBrazil"
};

static NSString * const LanguageSaveKey = @"kProtonMailCurrentLanguageKey";

#ifndef Enterprise
static NSString * const LanguageAppGroup = @"group.com.protonmail.protonmail";
#else
static NSString * const LanguageAppGroup = @"group.ch.protonmail.protonmail";
#endif

@implementation LanguageManager

+ (void)setupCurrentLanguage
{
    NSUserDefaults* shared = [[NSUserDefaults alloc] initWithSuiteName:LanguageAppGroup];
    NSString *currentLanguage = [shared objectForKey:LanguageSaveKey];
    if (!currentLanguage) {
        NSArray *languages = [shared objectForKey:@"AppleLanguages"];
        if (languages.count > 0) {
            currentLanguage = languages[0];
            [shared setObject:currentLanguage forKey:LanguageSaveKey];
            [shared synchronize];
        }
    }
#ifndef USE_ON_FLY_LOCALIZATION
    [shared setObject:@[currentLanguage] forKey:@"AppleLanguages"];
    [shared synchronize];
#else
    [NSBundle setLanguage:currentLanguage];
#endif
}

+ (NSArray *)languageStrings
{
    NSMutableArray *array = [NSMutableArray array];
    for (NSInteger i = 0; i < ELanguageCount; ++i) {
        [array addObject:NSLocalizedString(LanguageStrings[i], @"")];
    }
    return [array copy];
}

+ (NSString *)currentLanguageString
{
    NSString *string = @"";
    NSString *currentCode = [[[NSUserDefaults alloc] initWithSuiteName:LanguageAppGroup] objectForKey:LanguageSaveKey];
    for (NSInteger i = 0; i < ELanguageCount; ++i) {
        if ([currentCode isEqualToString:LanguageCodes[i]]) {
            string = NSLocalizedString(LanguageStrings[i], @"");
            break;
        }
    }
    return string;
}

+ (NSString *)currentLanguageCode
{
    return [[[NSUserDefaults alloc] initWithSuiteName:LanguageAppGroup] objectForKey:LanguageSaveKey];
}

+ (NSInteger)currentLanguageIndex
{
    NSInteger index = 0;
    NSString *currentCode = [[[NSUserDefaults alloc] initWithSuiteName:LanguageAppGroup] objectForKey:LanguageSaveKey];
    for (NSInteger i = 0; i < ELanguageCount; ++i) {
        if ([currentCode isEqualToString:LanguageCodes[i]]) {
            index = i;
            break;
        }
    }
    return index;
}


+ (ELanguage)currentLanguageEnum {
    NSInteger index = [self currentLanguageIndex];
    return (ELanguage)(index);
}

+ (void)saveLanguageByIndex:(NSInteger)index
{
    if (index >= 0 && index < ELanguageCount) {
        NSString *code = LanguageCodes[index];
        NSUserDefaults* shared = [[NSUserDefaults alloc] initWithSuiteName:LanguageAppGroup];
        [shared setObject:code forKey:LanguageSaveKey];
        [shared synchronize];
#ifdef USE_ON_FLY_LOCALIZATION
        [NSBundle setLanguage:code];
#endif
    }
}

+ (void)saveLanguageByCode:(NSString*)code {
    NSUserDefaults* shared = [[NSUserDefaults alloc] initWithSuiteName:LanguageAppGroup];
    [shared setObject:code forKey:LanguageSaveKey];
    [shared synchronize];
#ifdef USE_ON_FLY_LOCALIZATION
    [NSBundle setLanguage:code];
#endif
    
}


+ (BOOL)isCurrentLanguageRTL
{
    NSInteger currentLanguageIndex = [self currentLanguageIndex];
    return ([NSLocale characterDirectionForLanguage:LanguageCodes[currentLanguageIndex]] == NSLocaleLanguageDirectionRightToLeft);
}

@end

File: ProtonMail/ProtonMail/LanguageManager.swift
//
//  LanguageManager.swift
//  ProtonMail
//
//  Created by Yanfeng Zhang on 6/5/17.
//  Copyright © 2017 ProtonMail. All rights reserved.
//

import Foundation


//
//  LanguageManager.swift
//  ProtonMail
//
//  Created by Yanfeng Zhang on 6/5/17.
//  Copyright © 2017 ProtonMail. All rights reserved.
//

import Foundation

//
//public class LanguageManager {
//    
//    fileprivate static let kCurrentLanguageKey : String = "kProtonMailCurrentLanguageKey"
//    
//    //static NSString * const LanguageSaveKey = @"currentLanguageKey";
//    
//    //    class func currentLanguage() -> String {
//    //        if let currentLanguage = UserDefaults.standard.object(forKey: kCurrentLanguageKey) as? String {
//    //            return currentLanguage
//    //        }
//    //        return defaultLanguage()
//    //    }
//    
//    static func setupCurrentLanguage() {
//        var currentLanguage = UserDefaults.standard.object(forKey: kCurrentLanguageKey) as? String
//        if currentLanguage == nil {
//            if let languages =  UserDefaults.standard.object(forKey: "AppleLanguages") as? [String], languages.count > 0 {
//                currentLanguage = languages[0];
//                //                [[NSUserDefaults standardUserDefaults] setObject:currentLanguage forKey:kProtonMailCurrentLanguageKey];
//                //                [[NSUserDefaults standardUserDefaults] synchronize];
//            }
//        }
//        
//        
//        #if USE_ON_FLY_LOCALIZATION
//            Bundle.setLanguage(language: currentLanguage!)
//            //PMLog.D(aa);
//            //Bundle.setLanguage(currentLanguage)
//            //[NSBundle setLanguage:currentLanguage];
//            
//        #else
//            PMLog.D(aa);
//            //            [[NSUserDefaults standardUserDefaults] setObject:@[currentLanguage] forKey:@"AppleLanguages"];
//            //            [[NSUserDefaults standardUserDefaults] synchronize];
//        #endif
//    }
//    
//    //    let selectedLanguage = currentLanguage();
//    //    self.latestBundle = nil
//    //    Bundle.setLanguage(selectedLanguage)
//    //    UserDefaults.standard.set(selectedLanguage, forKey: kProtonMailCurrentLanguageKey)
//    //    UserDefaults.standard.synchronize()
//    
//    //+ (NSArray *)languageStrings;
//    //+ (NSString *)currentLanguageString;
//    //+ (void)saveLanguageByIndex:(NSInteger)index;
//    
//    
////    static func currentLanguageIndex() -> Int {
////        var index : Int = 0;
////        var currentCode = UserDefaults.standard.object(forKey: kCurrentLanguageKey) as? String
////        let allLanguage = SLItem.allItems()
////        
////        for l in allLanguage {
////            if currentCode == l {
////             
////            }
////        }
////        
////        
////        
////        NSString *currentCode = [[NSUserDefaults standardUserDefaults] objectForKey:kCurrentLanguageKey];
////        for (NSInteger i = 0; i < ELanguageCount; ++i) {
////            if ([currentCode isEqualToString:LanguageCodes[i]]) {
////                index = i;
////                break;
////            }
////        }
////        return index;
////    }
//    
//    
////    static func currentLanguageString() -> String {
////        var currentCode = UserDefaults.standard.object(forKey: kCurrentLanguageKey) as? String
////        let allLanguage = SLItem.allItems()
////        
////        for l in allLanguage {
////            if currentCode == l {
////                
////            }
////        }
////        NSString *currentCode = [[NSUserDefaults standardUserDefaults] objectForKey:kCurrentLanguageKey];
////        for (NSInteger i = 0; i < ELanguageCount; ++i) {
////            if ([currentCode isEqualToString:LanguageCodes[i]]) {
////                index = i;
////                break;
////            }
////        }
////        return index;
////    }
////    
//    
//    static func currentLanguageCode() -> String {
//        var currentLanguage = UserDefaults.standard.object(forKey: kCurrentLanguageKey) as? String
//        if currentLanguage == nil {
//            if let languages =  UserDefaults.standard.object(forKey: "AppleLanguages") as? [String], languages.count > 0 {
//                currentLanguage = languages[0];
////                [[NSUserDefaults standardUserDefaults] setObject:currentLanguage forKey:kProtonMailCurrentLanguageKey];
////                [[NSUserDefaults standardUserDefaults] synchronize];
//            }
//        }
//        return currentLanguage ?? ""
//    }
//
//
//    static func isCurrentLanguageRTL() -> Bool {
//        let out = NSLocale.characterDirection(forLanguage: self.currentLanguageCode())
//        return out == .rightToLeft
//    }
//    
//}
